Tests/ModelHelpers/AttributesTest.php

```diff
@@ -0,0 +1,189 @@
+	 * @return void
+	 */
+	public static function setUpBeforeClass()
+	{
+		static::$dataSets = array(
+			'banners'       => __DIR__ . '/Stubs/banners.csv',
+			'messages'      => __DIR__ . '/Stubs/messages.csv',
+			'users'         => __DIR__ . '/Stubs/users.csv',
+			'user_profiles' => __DIR__ . '/Stubs/user_profiles.csv'
+		);
+
+		parent::setUpBeforeClass();
+	}
+
+	/**
+	 * @covers \Joomla\Entity\ModelHelpers\Attributes::getAttribute()
+	 * @return void
+	 */
+	public function testGetAttribute()
+	{
+		$userModel = new User(self::$driver);
+
+		$user = $userModel->find(42);
+
+		$simpleAttribute = $user->getAttribute('id');
+
+		$mutatorAttribute = $user->getAttribute('newAccount');
+
+		$dateAttribute = $user->getAttribute('registerDate');
+
+		$castAttribute = $user->getAttribute('params');
+
+		$relationAttribute = $user->getAttribute('sentMessages');
+
+		$this->assertEquals(
+			42,
+			$simpleAttribute
+		);
+
+		$this->assertTrue(
+			$mutatorAttribute
+		);
+
+		$this->assertEquals(
+			'2010-02-13 00:34:42',
+			$dateAttribute
+		);
+
+		$this->assertEquals(
+			['test' => 'Object'],
+			$castAttribute
+		);
+
+		$this->assertInstanceOf(
+			Collection::class,
+			$relationAttribute
+		);
+
+		$this->assertInstanceOf(
+			Model::class,
+			$relationAttribute->first()
+		);
+
+		$this->assertEquals(
+			1,
+			$relationAttribute->first()->message_id
+		);
+
+		$this->expectException(AttributeNotFoundException::class);
+		$user->getAttribute("notExistent");
+	}
+
+	/**
+	 * @covers \Joomla\Entity\ModelHelpers\Attributes::getAttributeValue()
+	 * @return void
+	 */
+	public function testGetAttributeValue()
+	{
+		$userModel = new User(self::$driver);
+
+		$user = $userModel->find(42);
+
+		$simpleAttribute = $user->getAttributeValue('id');
+
+		$mutatorAttribute = $user->getAttributeValue('newAccount');
+
+		$dateAttribute = $user->getAttributeValue('registerDate');
+
+		$castAttribute = $user->getAttributeValue('params');
+
+		$this->expectException(AttributeNotFoundException::class);
+		$user->getAttributeValue('sentMessages');
+
+		$this->assertEquals(
+			42,
+			$simpleAttribute
+		);
+
+		$this->assertTrue(
+			$mutatorAttribute
+		);
+
+		$this->assertEquals(
+			'2010-02-13 00:34:42',
+			$dateAttribute
+		);
+
+		$this->assertEquals(
+			['test' => 'Object'],
+			$castAttribute
+		);
+	}
+
+	/**
+	 * @covers \Joomla\Entity\ModelHelpers\Attributes::setAttribute()
+	 * @return void
+	 */
+	public function testSetAttribute()
+	{
+		$userModel = new User(self::$driver);
+
+		$user = $userModel->find(42);
+
+		// Simple Attribute
+		$user->setAttribute('username', 'test');
+
+		// Mutator Attribute
+		$user->setAttribute('reset', 1);
+
+		// Date Attribute
+		$user->setAttribute('registerDate', '2010-02-13 00:34:43');
+
+		// Cast Attribute
+		$arr = ['test' => 'test'];
+		$user->setAttribute('params', $arr);
+
+		$this->expectException(AttributeNotFoundException::class);
```

Comment:
Same with this test

Reply:
I observed that expectException only has effect after it was called, but it makes it more clear to move them to another test anyway.

Reply:
Generally it's good practice to have a test per behaviour. Eventually we'll split these up even further. `testGetAttribute`, `testGetAttributeRetrievesDate`, `testGetAttributeWithMutator`, `testGetAttributeWithRelation`, `testGetAttributeNotFound` etc.

Reply:
fixed.